assignments/tdecacqu/notes-02.org

```diff
@@ -0,0 +1,561 @@
+(define (even? n)
+  (= (remainder n 2) 0))
+(define (fast-expt-iter b n a)
+  (cond ((= n 0) a)
+        ((even? n) (fast-expt-iter (* b b) (/ n 2) a))
+        (else (fast-expt-iter b (- n 1) (* a b)))))
+(define (fast-expt-iterative b n)
+  (fast-expt-iter b n 1))
+#+END_SRC
+
+*** Exercise 1.17
+
+Using halve and double, fast-mult can be implemented such as:
+a * b = 2 * a * b / 2   if b is even
+a * b = a + (a * (b - 1)) if b is odd
+
+#+BEGIN_SRC scheme :tangle 1.17.scm
+(define (even? n)
+  (= (remainder n 2) 0))
+(define (double x) (+ x x))
+(define (halve x) (/ x 2))
+(define (fast-mult a b)
+  (cond ((= b 0) 0)
+        ((= b 1) a)
+        ((even? b) (double (fast-mult a (halve b))))
+        (else (+ a (fast-mult a (- b 1))))))
+(fast-mult 3 5)
+#+END_SRC
+
+*** Exercise 1.18
+Using a counter, fast-mult can be written for an interative process:
+#+BEGIN_SRC scheme :tangle 1.18.scm
+(define (fast-mult-iter a b counter)
+  (cond ((= b 0) 0)
+        ((= b 1) (+ counter a))
+        ((even? b) (fast-mult-iter (double a) (halve b) counter))
+        (else (fast-mult-iter a (- b 1) (+ counter a)))))
+(define (fast-mult-iterative a b)
+    (fast-mult-iter a b 0))
+(fast-mult-iterative 5 9)
+#+END_SRC
+
+*** Exercise 1.19
+"Tpq transforms the pair (a,b) according to a = bq + aq + ap and b = bp + aq. Show that if we apply such a transformation Tpq twice, the effect is the same as using a single transformation Tp'q' of the same form"
+
+#+BEGIN_SRC text
+Let's apply Tpq one more time:
+a = (bp + aq) * q + (bq + aq + ap) * q + (bq + aq + ap) * p
+a = bpq + aq^2 + bq^2 + aq^2 + apq + bqp + aqp + ap^2
+a = b * (pq + q^2 + qp) + a * (pq + q^2 + qp) + a * (q^2 + p^2)
+
+b = (bp + aq) * p + (bq + aq + ap) * q
+b = bp^2 + aqp + bq^2 + aq^2 + apq
+b = b * (p^2 + q^2) + a * (qp + q^2 + pq)
+
+Thus Tp'q' is Tpq twice with:
+p' = p^2 + q^2
+q' = pq + q^2 + qp
+#+END_SRC
+
+#+BEGIN_SRC scheme :tangle 1.19.scm
+(define (fib n)
+  (fib-iter 1 0 0 1 n))
+(define (fib-iter a b p q count)
+  (cond ((= count 0) b)
+        ((even? count)
+         (fib-iter a
+                   b
+                   (+ (* p p) (* q q))
+                   (+ (* p q) (* q q) (* q p))
+                   (/ count 2)))
+        (else (fib-iter (+ (* b q) (* a q) (* a p))
+                        (+ (* b p) (* a q))
+                        p
+                        q
+                        (- count 1)))))
+#+END_SRC
+
+This feels like magic :)
+
+*** Exercise 1.20
+Compare gcd execution with normal-order and applicative-order evaluation.
+The question is how many time remainder is evaluated.
+#+BEGIN_SRC scheme :tangle 1.20.scm
+(define (gcd a b)
+  (if (= b 0)
+      a
+      (gcd b (remainder a b))))
+#+END_SRC
+
+Evaluation of (gcd 206 40):
+#+BEGIN_SRC text
+(gcd 40 (remainder 206 40))
+(gcd 6  (remainder 40 6))
+(gcd 4  (remainder 6 4))
+(gcd 2  (remainder 4 2))
+(gcd 2  0)
+#+END_SRC
+
+Remainder is called 4th time with normal order evaluation.
```

Comment:
Is this actually correct? I was thinking that with applicative order it does it 4 times, since the ``(remainder a b)`` is first expanded before calling gcd, while with normal order the ``(gcd ..)`` is first expanded and leads to the 18 evaluations?

Reply:
You are right, that was incorrect. For some reason I often got those two names mixed. Thanks for the correction.